src/tree.ts

```diff
@@ -48,7 +48,8 @@ export async function getOrderedListForProject(
     throw new Error(`Project ${project} not found`);
   }
   const upstream: Node[] = parentChainFromNode(node); // last element is the current node
-  const downstream: Node[] = childChainFromNode(node); // first element is the current node
+  const downstream = [node];
+  childChainFromNode(node, downstream); // first element is the current node
   return upstream.concat(downstream.slice(1)); // avoid duplication of current node
 }
```

Comment:
The change suggestion is close to this but I'm probably missing some TS refactoring, please add returned object type to the functions
```
async function getOrderedListForProject(
  definitionFileLocation: string,
  project: string,
  opts?: ReaderOpts
) : ¿PENDING? {
  const node = await getTreeForProject(definitionFileLocation, project, opts);
  if (!node) {
    throw new Error(`Project ${project} not found`);
  }
  return getOrderedList([tree]);
}

/**
 * Gets a plain node list of projects ordered by precedence
 * @param {string} node - The node to iterate
 */
async function getOrderedList(node: Node) {
  const finalLeaves = getFinalLeavesFromTree(node);
  return finalLeaves
    .map(leaf => parentChainFromNode(leaf))
    .sort((a, b) => b.length - a.length)
    .reduce((acc, chain) => {
      acc.push(
        ...chain.filter(c => !acc.map(e => e.project).includes(c.project))
      );
      return acc;
    }, []);
}

/**
 * It returns back the final leaves from a tree
 * @param {Array} nodes the array of nodes
 * @param {Array} latestLeavesFromTree
 * @param {Array} allreadyVisitedNodes
 */
function getFinalLeavesFromTree(
  nodes: Node[],
  latestLeavesFromTree = [],
  allreadyVisitedNodes = []
) {
  if (nodes && nodes.length > 0) {
    const allreadyVisitedPreviousNodes = [...allreadyVisitedNodes];
    const notIncludedChildren = nodes
      .filter(node => !allreadyVisitedPreviousNodes.includes(node.project))
      .reduce((acc, e) => {
        acc.push(
          ...e.children.filter(
            child => !acc.map(e => e.project).includes(child.project)
          )
        );
        return acc;
      }, []);
    allreadyVisitedNodes.push(...nodes.map(e => e.project));
    getFinalLeavesFromTree(
      notIncludedChildren,
      latestLeavesFromTree,
      allreadyVisitedNodes
    );

    latestLeavesFromTree.push(
      ...nodes.filter(
        node =>
          (!node.children || node.children.length === 0) &&
          !latestLeavesFromTree.map(e => e.project).includes(node.project)
      )
    );

    return latestLeavesFromTree;
  } else {
    return [];
  }
}
```

Reply:
I remember we had a conversation about this specifically few months ago @shubhbapna right?

Reply:
Yep and the issue was that i was getting different output then what we were predicting based of on reading the code. so in the end we decided that i will implement it based on what we want to do rather than translating the existing code